File: data/commands/lookcommand.js
function LookCommand() {
    Command.call(this);

    this.setDescription("Look at something (any object, character, or the current room).\n" +
                        "\n" +
                        "Examples: look, look at door, look earl, examine sign,\n" +
                        "          look at key in inventory");
}

LookCommand.prototype = new Command();
LookCommand.prototype.constructor = LookCommand;

LookCommand.prototype.execute = function(player, command) {

    this.prepareExecute(player, command);

    if (this.alias === "l" || this.alias === "look") {
        if (!this.hasWordsLeft()) {
            player.send(player.currentRoom.lookAtBy(player));
            return;
        }

        this.takeWord("at", Options.IfNotLast);
        if (!this.assertWordsLeft("Look at what?")) {
            return;
        }

        if (this.peekWord() === "inventory" || this.peekRest() === "in inventory") {
            player.execute("inventory");
            return;
        }
    } else {
        if (!this.assertWordsLeft("Examine what?")) {
            return;
        }
    }

    var object;
    var description = this.takeObjectsDescription();

    if (this.peekRest() === "in inventory") {
        object = this.objectByDescription(description, player.inventory);
        if (object) {
            player.send(object.lookAtBy(player));
        } else {
            player.send("You don't have that.");
        }
        return;
    }

    var room = this.currentRoom;

    var pool = room.characters.concat(room.items).concat(room.portals).concat(player.inventory);
    object = this.objectByDescription(description, pool);

    if (!object) {
        if (Util.isDirectionAbbreviation(description.name)) {
            description.name = Util.direction(description.name);
        }
        if (Util.isDirection(description.name)) {
            player.send("You look %1.".arg(description.name));
            player.direction = Util.vectorForDirection(description.name);
            player.lookAhead();
        } else if (description.name === "left") {
            player.send("You look to the left.");
            player.direction = [player.direction[1], -player.direction[0], player.direction[2]];
            player.lookAhead();
        } else if (description.name === "right") {
            player.send("You look to the right.");
            player.direction = [-player.direction[1], player.direction[0], player.direction[2]];
            player.lookAhead();
        } else if (description.name === "back" || description.name === "behind") {
            player.send("You look behind you.");
            player.direction = [-player.direction[0], -player.direction[1], -player.direction[2]];
            player.lookAhead();
        } else {
            player.send("Look where?");
        }
        return;
    }

    player.send(object.lookAtBy(player));

    function describeNearbyObject(nearbyObject, vector, position) {
        if (vector.angle(position) < Math.PI / 8) {
            var angle = Util.angleBetweenXYVectors(vector, position);

            if (nearbyObject.isPortal()) {
                player.send("On its %1 is %2.".arg(angle > 0 ? "right" : "left",
                                                   nearbyObject.nameWithDestinationFromRoom(room)));
            } else {
                player.send("On its %1 there's %2.".arg(angle > 0 ? "right" : "left",
                                                        nearbyObject.indefiniteName()));
            }
        }
    }

    var showNearbyObjects = (object.isItem() && !object.position.equals([0, 0, 0])) ||
                            object.isPortal();
    if (showNearbyObjects) {
        var vector = (object.isPortal() ? object.position.minus(room.position) : object.position);
        room.items.forEach(function(item) {
            if (item !== object) {
                describeNearbyObject(item, vector, item.position);
            }
        });
        room.portals.forEach(function(portal) {
            if (portal !== object) {
                describeNearbyObject(portal, vector, portal.position.minus(room.position));
            }
        });
    }

    if (object.isPortal() && object.canSeeThrough()) {
        var strength = room.eventMultiplier("Visual");
        if (player.weapon && player.weapon.name === "binocular") {
            strength *= 4;
        }

        var characters = VisualUtil.charactersVisibleThroughPortal(player, room, object, strength);
        var characterText = VisualUtil.describeCharactersRelativeTo(characters, player);
        if (characterText !== "") {
            player.send(characterText);
        }
    }
};

CommandRegistry.registerCommand("examine", new LookCommand());
CommandRegistry.registerCommand("l", new LookCommand());
CommandRegistry.registerCommand("look", new LookCommand());

File: src/commands/look_command.rs
use crate::actions;
use crate::entity::EntityRef;
use crate::entity::Realm;
use crate::player_output::PlayerOutput;
use crate::relative_direction::RelativeDirection;
use crate::utils::{direction_by_abbreviation, is_direction, vector_for_direction};

use super::{inventory_command::inventory, CommandHelpers, CommandLineProcessor};

/// Makes the character look at *something*.
pub fn look(
    realm: &mut Realm,
    player_ref: EntityRef,
    helpers: CommandHelpers,
) -> Result<Vec<PlayerOutput>, String> {
    let (player, room) = realm.character_and_room_res(player_ref)?;

    let processor = helpers.command_line_processor;

    let alias = processor.take_word().unwrap();
    if alias.starts_with('l') {
        if !processor.has_words_left() {
            return actions::look_at_entity(realm, player_ref, player.current_room());
        }

        processor.skip_connecting_word("at");
    }

    if processor.peek_rest() == "inventory" || processor.peek_rest() == "in inventory" {
        return inventory(
            realm,
            player_ref,
            CommandHelpers {
                command_line_processor: &mut CommandLineProcessor::new(player_ref, "inventory"),
                ..helpers
            },
        );
    }

    let description = processor.take_entity_description().ok_or_else(|| {
        if alias == "examine" {
            "Examine what?".to_owned()
        } else {
            "Look at what?".to_owned()
        }
    })?;

    if processor.peek_rest() == "in inventory" {
        let entity_ref = processor
            .entity_by_description(realm, player.inventory(), description)
            .ok_or("You don't have that.")?;
        return actions::look_at_entity(realm, player_ref, entity_ref);
    }

    let pool = [
        &room.characters()[..],
        &room.items()[..],
        &room.portals()[..],
        &player.inventory()[..],
    ]
    .concat();

    let target = description.name.as_ref();
    let maybe_entity = processor
        .entity_by_description(realm, &pool, description.clone())
        .and_then(|entity_ref| realm.entity(entity_ref));
    let output = match maybe_entity {
        Some(entity) => {
            let mut output = Vec::<PlayerOutput>::new();
            if is_direction(target) {
                push_output_string!(output, player_ref, format!("You look {}.\n", target));
            } else if let Some(direction) = direction_by_abbreviation(target) {
                push_output_string!(output, player_ref, format!("You look {}.\n", direction));
            }
            output.append(&mut actions::look_at_entity(
                realm,
                player_ref,
                entity.entity_ref(),
            )?);
            output
        }
        None => {
            let (output, direction) = if let Some(direction) = direction_by_abbreviation(target) {
                let vector = vector_for_direction(direction).unwrap();
                (format!("You look {}.\n", direction), vector)
            } else if let Some(direction) = vector_for_direction(target) {
                (format!("You look {}.\n", target), direction)
            } else if let Some(relative_direction) = RelativeDirection::from_string(target) {
                (
                    match relative_direction {
                        RelativeDirection::Ahead => "You look ahead.\n",
                        RelativeDirection::Left => "You look to the left.\n",
                        RelativeDirection::Right => "You look to the right.\n",
                        RelativeDirection::Behind => "You look behind you.\n",
                        RelativeDirection::Up => "You look up.\n",
                        RelativeDirection::Down => "You look down.\n",
                    }
                    .to_owned(),
                    relative_direction.from(player.direction()),
                )
            } else {
                return Err(if alias == "examine" {
                    "Examine what?".to_owned()
                } else {
                    "Look where?".to_owned()
                });
            };

            let mut output = vec![PlayerOutput::new_from_string(player_ref.id(), output)];
            output.append(&mut actions::look_in_direction(
                realm, player_ref, &direction,
            )?);
            output
        }
    };

    if let Some(direction) = vector_for_direction(target) {
        actions::change_direction(realm, player_ref, direction);
    }

    Ok(output)
}
